File: server/app/utils/validation.py
# server/app/utils/validation.py
# Enhanced validation utilities

import re
from bson import ObjectId
from functools import wraps
from flask import request
from .error_handler import ValidationError
import mimetypes
import os

def validate_objectid(value, field_name="ID"):
    """Validate MongoDB ObjectId"""
    if not value:
        raise ValidationError(f"{field_name} is required")
    
    if not ObjectId.is_valid(value):
        raise ValidationError(f"Invalid {field_name} format")
    
    return ObjectId(value)

def validate_email(email):
    """Validate email format"""
    if not email:
        raise ValidationError("Email is required")
    
    email_pattern = r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"
    if not re.match(email_pattern, email):
        raise ValidationError("Invalid email format")
    
    return email.lower().strip()

def validate_password(password):
    """Validate password strength"""
    if not password:
        raise ValidationError("Password is required")
    
    min_length = int(os.environ.get('PASSWORD_MIN_LENGTH', 8))
    
    if len(password) < min_length:
        raise ValidationError(f"Password must be at least {min_length} characters long")
    
    if not re.search(r'[a-z]', password):
        raise ValidationError("Password must contain at least one lowercase letter")
    
    if not re.search(r'[A-Z]', password):
        raise ValidationError("Password must contain at least one uppercase letter")
    
    if not re.search(r'\d', password):
        raise ValidationError("Password must contain at least one number")
    
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        raise ValidationError("Password must contain at least one special character")
    
    return password

def validate_file_upload(file, allowed_extensions=None, max_size_mb=50):
    """Validate file upload"""
    if not file:
        raise ValidationError("No file provided")
    
    if file.filename == '':
        raise ValidationError("No file selected")
    
    # Check file extension
    if allowed_extensions:
        file_ext = file.filename.rsplit('.', 1)[1].lower() if '.' in file.filename else ''
        if file_ext not in allowed_extensions:
            raise ValidationError(f"File type not allowed. Allowed types: {', '.join(allowed_extensions)}")
    
    # Check file size
    file.seek(0, os.SEEK_END)
    file_size = file.tell()
    file.seek(0)  # Reset file pointer
    
    max_size_bytes = max_size_mb * 1024 * 1024
    if file_size > max_size_bytes:
        raise ValidationError(f"File size exceeds {max_size_mb}MB limit")
    
    # Validate MIME type
    mime_type, _ = mimetypes.guess_type(file.filename)
    if allowed_extensions and mime_type:
        allowed_mimes = {
            'pdf': 'application/pdf',
            'doc': 'application/msword',
            'docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            'ppt': 'application/vnd.ms-powerpoint',
            'pptx': 'application/vnd.openxmlformats-officedocument.presentationml.presentation'
        }
        
        expected_mime = None
        for ext in allowed_extensions:
            if ext in allowed_mimes and file.filename.lower().endswith(f'.{ext}'):
                expected_mime = allowed_mimes[ext]
                break
        
        if expected_mime and mime_type != expected_mime:
            raise ValidationError(f"File content does not match extension")
    
    return file

def validate_pagination(page=None, limit=None, max_limit=100):
    """Validate pagination parameters"""
    try:
        page = int(page) if page else 1
        limit = int(limit) if limit else 20
    except (ValueError, TypeError):
        raise ValidationError("Page and limit must be integers")
    
    if page < 1:
        raise ValidationError("Page must be greater than 0")
    
    if limit < 1:
        raise ValidationError("Limit must be greater than 0")
    
    if limit > max_limit:
        raise ValidationError(f"Limit cannot exceed {max_limit}")
    
    return page, limit

def validate_search_query(query, min_length=2, max_length=200):
    """Validate search query"""
    if not query:
        raise ValidationError("Search query is required")
    
    query = query.strip()
    
    if len(query) < min_length:
        raise ValidationError(f"Search query must be at least {min_length} characters")
    
    if len(query) > max_length:
        raise ValidationError(f"Search query cannot exceed {max_length} characters")
    
    # Prevent potentially malicious queries
    dangerous_patterns = [
        r'\$',  # MongoDB injection
        r'<script',  # XSS
        r'javascript:',  # XSS
        r'eval\(',  # Code injection
    ]
    
    for pattern in dangerous_patterns:
        if re.search(pattern, query, re.IGNORECASE):
            raise ValidationError("Invalid characters in search query")
    
    return query

def validate_json_body(required_fields=None, optional_fields=None):
    """Decorator to validate JSON request body"""
    def decorator(f):
        @wraps(f)
        def decorated_function(*args, **kwargs):
            if not request.is_json:
                raise ValidationError("Request must be JSON")
            
            data = request.get_json()
            if not data:
                raise ValidationError("Request body is required")
            
            # Check required fields
            if required_fields:
                for field in required_fields:
                    if field not in data or data[field] is None:
                        raise ValidationError(f"Field '{field}' is required")
            
            # Check for unexpected fields
            if optional_fields is not None:
                allowed_fields = set(required_fields or []) | set(optional_fields or [])
                unexpected_fields = set(data.keys()) - allowed_fields
                if unexpected_fields:
                    raise ValidationError(f"Unexpected fields: {', '.join(unexpected_fields)}")
            
            return f(*args, **kwargs)
        return decorated_function
    return decorator

def validate_rate_limit_key(key):
    """Validate rate limiting key"""
    if not key:
        raise ValidationError("Rate limit key is required")
    
    # Sanitize key to prevent injection
    sanitized_key = re.sub(r'[^a-zA-Z0-9:._-]', '', key)
    if len(sanitized_key) != len(key):
        raise ValidationError("Invalid characters in rate limit key")
    
    return sanitized_key

class QueryValidator:
    """Helper class for validating database queries"""
    
    @staticmethod
    def validate_sort_field(field, allowed_fields):
        """Validate sort field"""
        if field not in allowed_fields:
            raise ValidationError(f"Invalid sort field. Allowed fields: {', '.join(allowed_fields)}")
        return field
    
    @staticmethod
    def validate_filter_value(value, field_type='string'):
        """Validate filter value based on type"""
        if field_type == 'objectid':
            return validate_objectid(value)
        elif field_type == 'email':
            return validate_email(value)
        elif field_type == 'string':
            if not isinstance(value, str):
                raise ValidationError(f"Filter value must be a string")
            return value.strip()
        elif field_type == 'int':
            try:
                return int(value)
            except (ValueError, TypeError):
                raise ValidationError(f"Filter value must be an integer")
        elif field_type == 'bool':
            if isinstance(value, bool):
                return value
            if isinstance(value, str):
                return value.lower() in ('true', '1', 'yes', 'on')
            raise ValidationError(f"Filter value must be a boolean")
        else:
            return value

File: backend/middlewares/validation.js
// backend/middlewares/validation.js
// Express validation middleware

const mongoose = require('mongoose');

// Custom error classes
class ValidationError extends Error {
  constructor(message, details = null) {
    super(message);
    this.name = 'ValidationError';
    this.statusCode = 400;
    this.code = 'VALIDATION_ERROR';
    this.details = details;
  }
}

// Validate ObjectId
const validateObjectId = (value, fieldName = 'ID') => {
  if (!value) {
    throw new ValidationError(`${fieldName} is required`);
  }
  
  if (!mongoose.Types.ObjectId.isValid(value)) {
    throw new ValidationError(`Invalid ${fieldName} format`);
  }
  
  return new mongoose.Types.ObjectId(value);
};

// Validate pagination parameters
const validatePagination = (page = 1, limit = 20, maxLimit = 100) => {
  const pageNum = parseInt(page);
  const limitNum = parseInt(limit);
  
  if (isNaN(pageNum) || pageNum < 1) {
    throw new ValidationError('Page must be a positive integer');
  }
  
  if (isNaN(limitNum) || limitNum < 1) {
    throw new ValidationError('Limit must be a positive integer');
  }
  
  if (limitNum > maxLimit) {
    throw new ValidationError(`Limit cannot exceed ${maxLimit}`);
  }
  
  return { page: pageNum, limit: limitNum };
};

// Validate JSON body middleware
const validateJsonBody = (requiredFields = [], optionalFields = null) => {
  return (req, res, next) => {
    try {
      if (!req.body || Object.keys(req.body).length === 0) {
        throw new ValidationError('Request body is required');
      }
      
      // Check required fields
      for (const field of requiredFields) {
        if (!(field in req.body) || req.body[field] === null || req.body[field] === undefined) {
          throw new ValidationError(`Field '${field}' is required`);
        }
      }
      
      // Check for unexpected fields if optionalFields is specified
      if (optionalFields !== null) {
        const allowedFields = new Set([...requiredFields, ...optionalFields]);
        const bodyFields = new Set(Object.keys(req.body));
        const unexpectedFields = [...bodyFields].filter(field => !allowedFields.has(field));
        
        if (unexpectedFields.length > 0) {
          throw new ValidationError(`Unexpected fields: ${unexpectedFields.join(', ')}`);
        }
      }
      
      next();
    } catch (error) {
      next(error);
    }
  };
};

// Validate query parameters
const validateQuery = (allowedParams = []) => {
  return (req, res, next) => {
    try {
      const queryKeys = Object.keys(req.query);
      const unexpectedParams = queryKeys.filter(key => !allowedParams.includes(key));
      
      if (unexpectedParams.length > 0) {
        throw new ValidationError(`Unexpected query parameters: ${unexpectedParams.join(', ')}`);
      }
      
      next();
    } catch (error) {
      next(error);
    }
  };
};

// Validate room data
const validateRoomData = (req, res, next) => {
  try {
    const { name, unit_code, topic, max_participants } = req.body;
    
    // Validate name
    if (!name || typeof name !== 'string' || name.trim().length < 3) {
      throw new ValidationError('Room name must be at least 3 characters long');
    }
    
    if (name.length > 100) {
      throw new ValidationError('Room name cannot exceed 100 characters');
    }
    
    // Validate unit_code
    if (!unit_code || typeof unit_code !== 'string' || unit_code.trim().length < 2) {
      throw new ValidationError('Unit code must be at least 2 characters long');
    }
    
    if (unit_code.length > 20) {
      throw new ValidationError('Unit code cannot exceed 20 characters');
    }
    
    // Validate topic
    if (!topic || typeof topic !== 'string' || topic.trim().length < 3) {
      throw new ValidationError('Topic must be at least 3 characters long');
    }
    
    if (topic.length > 200) {
      throw new ValidationError('Topic cannot exceed 200 characters');
    }
    
    // Validate max_participants
    if (max_participants !== undefined) {
      const maxPart = parseInt(max_participants);
      if (isNaN(maxPart) || maxPart < 2 || maxPart > 100) {
        throw new ValidationError('Max participants must be between 2 and 100');
      }
    }
    
    // Sanitize data
    req.body.name = name.trim();
    req.body.unit_code = unit_code.trim().toUpperCase();
    req.body.topic = topic.trim();
    
    next();
  } catch (error) {
    next(error);
  }
};

// Validate message data
const validateMessageData = (req, res, next) => {
  try {
    const { content, type = 'text' } = req.body;
    
    // Validate content
    if (!content || typeof content !== 'string' || content.trim().length === 0) {
      throw new ValidationError('Message content is required');
    }
    
    if (content.length > 1000) {
      throw new ValidationError('Message content cannot exceed 1000 characters');
    }
    
    // Validate type
    const allowedTypes = ['text', 'resource', 'announcement', 'poll', 'timer', 'system'];
    if (!allowedTypes.includes(type)) {
      throw new ValidationError(`Invalid message type. Allowed types: ${allowedTypes.join(', ')}`);
    }
    
    // Sanitize content
    req.body.content = content.trim();
    
    next();
  } catch (error) {
    next(error);
  }
};

// Validate file upload
const validateFileUpload = (allowedMimeTypes = [], maxSizeMB = 50) => {
  return (req, res, next) => {
    try {
      if (!req.file && !req.files) {
        throw new ValidationError('No file uploaded');
      }
      
      const file = req.file || (req.files && req.files[0]);
      
      if (!file) {
        throw new ValidationError('No file found in request');
      }
      
      // Check MIME type
      if (allowedMimeTypes.length > 0 && !allowedMimeTypes.includes(file.mimetype)) {
        throw new ValidationError(`Invalid file type. Allowed types: ${allowedMimeTypes.join(', ')}`);
      }
      
      // Check file size
      const maxSizeBytes = maxSizeMB * 1024 * 1024;
      if (file.size > maxSizeBytes) {
        throw new ValidationError(`File size exceeds ${maxSizeMB}MB limit`);
      }
      
      next();
    } catch (error) {
      next(error);
    }
  };
};

// Rate limiting validation
const validateRateLimit = (windowMs = 15 * 60 * 1000, maxRequests = 100) => {
  const requests = new Map();
  
  return (req, res, next) => {
    try {
      const key = req.ip || req.connection.remoteAddress;
      const now = Date.now();
      const windowStart = now - windowMs;
      
      // Clean old entries
      if (requests.has(key)) {
        const userRequests = requests.get(key).filter(time => time > windowStart);
        requests.set(key, userRequests);
      }
      
      const currentRequests = requests.get(key) || [];
      
      if (currentRequests.length >= maxRequests) {
        throw new ValidationError('Too many requests, please try again later', 429);
      }
      
      currentRequests.push(now);
      requests.set(key, currentRequests);
      
      next();
    } catch (error) {
      next(error);
    }
  };
};

module.exports = {
  validateObjectId,
  validatePagination,
  validateJsonBody,
  validateQuery,
  validateRoomData,
  validateMessageData,
  validateFileUpload,
  validateRateLimit
};